docs/adrs/0000-composite-actions.md

```diff
@@ -0,0 +1,94 @@
+- We are free to bump this limit in the future, the code will be written to just require updating a variable. If the graph evaluates beyond the recursion limit, the job will fail in the pre-job phase (The `Set up job` step).
+- A composite actions interface is its inputs and outputs, nothing else is carried over when invoking recursively.
+
+### Pre/Post Steps in nested Actions
+
+- We do not plan on adding the ability to configure a customizable pre or post step for composite actions at this time. However, we will execute the pre and post steps of any actions referenced in a composite action.
+- Composite actions will generate a single pre-step and post-step for the entire composite action, even if there are multiple pre-steps and post-steps in the referenced actions.
+  - These steps will execute following the same ordering rules we have today, first to run has their pre step run first and their post step run last.
+  - For example, if you had a composite action with two pre steps and two posts steps:
+
+  ```
+  - uses: action1
+  - uses: composite1
+  - uses: action2
+  ```
+
+  The order of execution would be:
+
+  ```
+  - prestep-action1
+  - prestep-composite1
+    - prestep-composite1-first-action-referenced
+    - prestep-composite1-second-action-referenced
+  - prestep-action2
+  - the job steps
+  - poststep-action2
+  - poststep-composite1
+    - poststep-composite1-the-second-action-referenced
+    - poststep-composite1-first-action-referenced
+  - poststep-action1
+  ```
+
+#### Set-state
+
+- While the composite action has an individual combined pre/post action, the `set-state` command will not be shared.
+- If the `set-state` command is used during a composite step, only the action that originally called `set-state` will have access to the env variable during the post run step.
+  - This prevents multiple actions that set the same state from interfering with the execution of another action's post step.
+
+### Resolve Action Endpoint changes
+
+- The resolve actions endpoint will now validate policy to ensure that the given workflow run has access to download that action.
+  - It will return a value indicated security was checked. Older server versions would lack this boolean. For composite actions that try to resolve an action, if the server does not respond with this value, we will fail the job as we could violate access policy if we allowed the composite action to continue.
+    - Oder GHES/GHAE customers with newer runners will be locked out of composite uses steps until they upgrade their instance.
+
+### Local actions
+- TODO describe this more
+- We should support pre and post actions for local actions
+
+
+### If, continue-on-error, timeout-minutes - Not being considered at this time
+
+- `if`, `continue-on-error`, `timeout-minutes` could be supported in composite run/uses steps. These values were not originally supported in our composite run steps implementation.
```

Comment:
Might be good to prioritize the `if` keyword sooner rather than later. Feedback from @david-gang

> I see that `if` is not planned to be supported. This could be helpful for us. for example if i have two workflows which are nearly identical but have a big chunk of identical steps. part of the steps contain if clauses. Now i can't unite them in one action and call this action from both workflows. If the if would be supported i could do it. I know that i can use a javascript action but why would i need javascript if it is just a number of bash steps.

Reply:
This would be amazing feature https://github.com/prince-chrismc/jwt-cpp/runs/3448042058